### PR TITLE
CMake Refactor

### LINKED ISSUE #56: Build system refactoring
[ 75%] Performing build step for 'viennals_external'
gmake[4]: *** No rule to make target 'buildDependencies'.  Stop.
make[3]: *** [external/upstream/viennals/CMakeFiles/viennals_external.dir/build.make:86: ../dependencies/Stamp/viennals_external/viennals_external-build] Error 2
make[2]: *** [CMakeFiles/Makefile2:187: external/upstream/viennals/CMakeFiles/viennals_external.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:142: CMakeFiles/buildDependencies.dir/rule] Error 2
make: *** [Makefile:169: buildDependencies] Error 2

when i make dependencies,there is a problem.How can I figure out this?

_Originally posted by @Yanthirteen13 in https://github.com/ViennaTools/ViennaPS/issues/50#issuecomment-1846634482_
           

We are presently undergoing a significant overhaul to enhance the entire build system in ViennaPS/ViennaLS. As a result, there may be occasional build issues in ViennaPS during this transitional phase.

To address this temporarily, we recommend utilizing an earlier version of both libraries.

We are working diligently to complete the refactoring process promptly, ensuring that the library can be seamlessly utilized without any issues in the near future. We appreciate your patience and understanding during this improvement period.

### PR DESCRIPTION
**Pull request by Curve**
This Pull Request modernizes CMake (resolves #56):
 - Use CPM / PackageProject

 - Update Include Structure
   - Don't artificially flatten
   - Use q-char-sequences where appropriate

 - Copy required shared libraries when required / Setup proper path automatically when required

- Modernize Python Bindings (#65)
   - Use scikit-build-core
   - Re-use the shared libraries contained in the VTK-Python package
     - [x] On Linux _(using RPATH)_ [✅ Tested]
     - [x] On MacOS _(using RPATH)_ [✅ Tested]
     - [x] On Windows _(`os.add_dll_directory`)_ [⏳ Via `viennaps.libs` folder, awaiting upstream changes]

- Add library switch to ease location of installed dependencies (ViennaLS, ViennaRay, ...)

### CODE REVIEW

**Comment by Curve** [comment]
Note to self: I still have to adjust the workflows

File: include/viennaps/cellSet/csSurfaceCellSet.hpp [hunk 1]
```diff
@@ -56,15 +56,16 @@ template <class T, int D, class cellType = cellBase<T>> class csSurfaceCellSet {
       }
     }
 
-    std::vector<cellType &> getNeighbors(const cellType &cell) {
-      std::vector<cellType &> neighbors;
-      auto cellSet = domain->getCellSet();
+    // What happened here?
```
Comment (Curve) [inline include/viennaps/cellSet/csSurfaceCellSet.hpp:59]: @tobre1 Was this some sort of merge conflict on your side?

**Comment by Curve** [comment]
This PR should now be done. We should still wait for the bindings workflow to finish (https://github.com/ViennaTools/ViennaPS/actions/runs/8315558071) to confirm that the bindings work properly on windows though :)

**Comment by Curve** [comment]
> This PR should now be done. We should still wait for the [bindings workflow](https://github.com/ViennaTools/ViennaPS/actions/runs/8315558071) to finish to confirm that the bindings work properly on windows though :)

For some god forsaken reason the python vtk package dlls on windows are suffixed with some random hashes:

image

I'll have to see what I can do here...

**Comment by Curve** [comment]
I've evaluated some possible solutions but most are way too hacky to be deemed viable - I have thus raised an issue on VTKs side: https://gitlab.kitware.com/vtk/vtk/-/issues/19279.

For now I'm going to add a `viennaps.libs` folder on those targets that require it _(currently only windows)_ and copied the VTK libraries there for the time being.

**Comment by Curve** [comment]
The VTK libraries are now installed in the `viennaps.libs` folder only on windows for the time being - In case the VTK issue is resolved we could remove the libs folder on windows as well :)

**Comment by tobre1** [comment]
When building the examples, the binaries are now placed in `build/examples-bin/`. This change should also be addressed in the README Basic Examples - Building section.

**Comment by tobre1** [comment]
When local ViennaLS/ViennaRay packages are used (specified in VIENNAPS_LOOKUP_DIRS), OpenMP is not added to the dependencies when building the examples.  I get the following error:
```
[cmake] CMake Error at examples/GDSReader/CMakeLists.txt:3 (add_executable):
[cmake]   Target "GDSReader" links to target "OpenMP::OpenMP_CXX" but the target was
[cmake]   not found.  Perhaps a find_package() call is missing for an IMPORTED
[cmake]   target, or an ALIAS target is missing?
```

**Comment by Curve** [comment]
> When building the examples, the binaries are now placed in `build/examples-bin/`. This change should also be addressed in the README Basic Examples - Building section.

> When local ViennaLS/ViennaRay packages are used (specified in VIENNAPS_LOOKUP_DIRS), OpenMP is not added to the dependencies when building the examples. I get the following error:
> 
> ```
> [cmake] CMake Error at examples/GDSReader/CMakeLists.txt:3 (add_executable):
> [cmake]   Target "GDSReader" links to target "OpenMP::OpenMP_CXX" but the target was
> [cmake]   not found.  Perhaps a find_package() call is missing for an IMPORTED
> [cmake]   target, or an ALIAS target is missing?
> ```

Thanks for the thorough testing!

I've updated the README and pushed a fix for ViennaRay (regarding OpenMP) here: https://github.com/ViennaTools/ViennaRay/pull/48

Please try to install ViennaRay locally again with commit `6f2d980`

**Comment by Curve** [comment]
> I've evaluated some possible solutions but most are way too hacky to be deemed viable - I have thus raised an issue on VTKs side: https://gitlab.kitware.com/vtk/vtk/-/issues/19279.
> 
> For now I'm going to add a `viennaps.libs` folder on those targets that require it _(currently only windows)_ and copy the VTK libraries there for the time being.

VTK is working on a new way to make this possible, however it all seems to be highly work in progress right now

**Pull request merged**